routing/payment_state_machine.go

```diff
@@ -0,0 +1,419 @@
+	// by the payment hash of the active payments.
+	paymentsBucketKey = []byte("router-payment-state-machine")
+
+	// paymentStateKey is a key used in an active payment bucket to access
+	// the current payment state.
+	paymentStateKey = []byte("payment-state-key")
+
+	// paymentAttemptKey is used to access the information of the current
+	// payment attempt sent to the switch. This key is only populated when
+	// the payment is in the state paymentStateAttemptSent.
+	paymentAttemptKey = []byte("payment-attempt-key")
+
+	// paymentErrorKey is a key that accesses the last encountered error
+	// for payment attempts for the active payment. It is populated when
+	// the active payment is in the state paymentStateFailed, and sometimes
+	// in other states if the previous payment attempt failed.
+	paymentErrorKey = []byte("payment-error-key")
+
+	// ErrNotFound is an error returned if the given payment hash was not
+	// found.
+	ErrNotFound = errors.New("payment hash not found")
+)
+
+// paymentStateMachine is the persistent state machine used by the
+// ChannelRouter to track the progress of active payments. It is used to keep
+// information across restarts, needed to properly retry failed paymets, and
+// handle payment errors and successes encountered for a payment sent before
+// the ChannelRouter was restarted.
+type paymentStateMachine struct {
+	db *channeldb.DB
+}
+
+// paymentState is the type used to indicate which state a given payment has in
+// the payment state machine.
+type paymentState byte
+
+const (
+	// paymentStateIntended is the starting state of a payment. A payment
+	// in this state is intended for sending, but haven't yet had a
+	// paymentAttempt sent to the Switch..
+	paymentStateIntended paymentState = 1
+
+	// paymentStateAttemptSent is the state a payment enters after we have
+	// crafted a paymentAttempt for it and sent it to the Switch. In this
+	// state we expect to find the paymentAttempt in the DB, such that we
+	// have the information necessary to handle any results received from
+	// the Switch. After a restart the paymentAttempt can poentially have
+	// been lost by the Switch, in case we can go back to the initial state
+	// and craft a new paymentAttempt to retry.
+	paymentStateAttemptSent paymentState = 2
+
+	// paymentStateSuccess is the state a payment enters after it has
+	// completed successfully. In this state we expect to find the
+	// completed OutgoingPayment in the payments DB. The payment is kept in
+	// this state by the payment state machine to avoid mistakenly
+	// resending a payment to the same payment hash.
+	paymentStateSuccess paymentState = 3
+
+	// paymentStateFailed is the state a payment enters when it has failed
+	// permanently. In this state we will find the error encounterd in the
+	// DB. In this state it is safe to attempt to resend a payment to the
+	// same payment hash.
+	paymentStateFailed paymentState = 4
+
+	// TODO(halseth): timeout/cancel state?
```

Comment:
Yeah would be nice to cancel eventually. Timeout could either be an on-chain timeout, or a payment attempt timeout.

Reply:
Should be an easy addition, by just adding them as terminal states.